linker/src/lib.rs

```diff
@@ -1,27 +1,88 @@
+use std::{collections::BTreeSet, iter::once};
+
 use ast::{
     object::{Location, PILGraph},
-    parsed::{Expression, PILFile, PilStatement},
+    parsed::{
+        build::{direct_reference, namespaced_reference},
+        Expression, PILFile, PilStatement, SelectedExpressions,
+    },
 };
 use number::FieldElement;
 
 /// a monolithic linker which outputs a single AIR
 pub fn link<T: FieldElement>(mut graph: PILGraph<T>) -> PILFile<T> {
-    assert_eq!(graph.objects.len(), 1, "only one machine is supported");
-    let object = graph
-        .objects
-        .remove(&Location::default().join("main"))
-        .unwrap();
-    let location = Location::default().join("main");
+    let mut queue = BTreeSet::from([Location::default().join("main")]);
```

Comment:
Why a BTreeSet? What's the ordering that this is relying on in the BFS below?

Reply:
Ah now I remember. There can be more than one link to a given submachine (because the submachine can have many functions). Using a set makes sure that each machine is processed exactly once. I didn't really care about the ordering.

Reply:
But now I realise if I don't care about the ordering I could just iterate through the original `BTreeMap`...

Reply:
BTreeSet used for deduplication.